File: src-tauri/src/spark.rs
//! Tier 1.5 「靈感火花」：對單一概念呼叫本地 Ollama 生成「延伸方向 + 隨想」。
//!
//! 角色 ≠ 被砍的 Tier 1（連結過濾器）。火花是發散式激發靈感,純建議、拋棄式——
//! 輸出不入庫,人工挑/改後才由前端走 add_thought 升格成新 thought（守紅線 #2）。
//!
//! spike 實證（workspace/sandbox/qwen3-spark-spike）：qwen3 家族守繁中、守格式、會發散;
//! qwen2.5 守不住格式。預設 qwen3.5:4b（6GB 裝得下）,8GB+ 可選 qwen3:8b。

use std::time::Duration;

const OLLAMA: &str = "http://localhost:11434";

/// 火花結果：三條延伸方向 + 一段隨想。
#[derive(Debug, serde::Serialize)]
pub struct SparkResult {
    pub directions: Vec<String>,
    pub musing: String,
}

/// Ollama 健康狀態,供前端優雅降級。
#[derive(Debug, serde::Serialize)]
pub struct SparkHealth {
    pub ollama_up: bool,
    pub model_ready: bool,
}

#[derive(serde::Serialize)]
struct GenReq<'a> {
    model: &'a str,
    prompt: String,
    stream: bool,
    // qwen3 thinking mode 預設開,會吐 <think> 污染輸出 → 必關。
    think: bool,
    options: GenOpts,
}

#[derive(serde::Serialize)]
struct GenOpts {
    temperature: f32,
}

#[derive(serde::Deserialize)]
struct GenResp {
    response: String,
}

#[derive(serde::Deserialize)]
struct TagsResp {
    models: Vec<TagModel>,
}

#[derive(serde::Deserialize)]
struct TagModel {
    name: String,
}

/// spike 定案模板：去 `<>` 佔位 + 要求台灣慣用詞彙（壓制 4b 偶帶的陸式詞彙）。
pub fn build_prompt(thought: &str) -> String {
    format!(
        "你是發想助手。針對下面這個想法,給我激發靈感的延伸,不要完整方案,要發散、可以歪。\n\
         務必用繁體中文與台灣慣用詞彙。嚴格照以下格式輸出,不要加任何其他文字:\n\n\
         方向:\n\
         - 關鍵詞或切角一\n\
         - 關鍵詞或切角二\n\
         - 關鍵詞或切角三\n\
         隨想:\n\
         (50 字以內,一段聯想短文)\n\n\
         想法:「{thought}」"
    )
}

/// 對單一概念生成火花。
pub async fn generate(model: &str, thought: &str, temperature: f32) -> Result<SparkResult, String> {
    let client = reqwest::Client::builder()
        .timeout(Duration::from_secs(60))
        .build()
        .map_err(|e| e.to_string())?;
    let req = GenReq {
        model,
        prompt: build_prompt(thought),
        stream: false,
        think: false,
        options: GenOpts { temperature },
    };
    let resp = client
        .post(format!("{OLLAMA}/api/generate"))
        .json(&req)
        .send()
        .await
        .map_err(|e| friendly_err(&e))?;
    if !resp.status().is_success() {
        return Err(format!("Ollama 回應 {}", resp.status()));
    }
    let body: GenResp = resp.json().await.map_err(|e| e.to_string())?;
    Ok(parse_spark(&body.response))
}

/// 探活 + 確認目標模型已拉,供前端決定火花入口要不要亮。
pub async fn health(model: &str) -> SparkHealth {
    let down = SparkHealth {
        ollama_up: false,
        model_ready: false,
    };
    let client = match reqwest::Client::builder()
        .timeout(Duration::from_secs(3))
        .build()
    {
        Ok(c) => c,
        Err(_) => return down,
    };
    match client.get(format!("{OLLAMA}/api/tags")).send().await {
        Ok(r) if r.status().is_success() => match r.json::<TagsResp>().await {
            Ok(tags) => SparkHealth {
                ollama_up: true,
                model_ready: tags.models.iter().any(|m| m.name == model),
            },
            Err(_) => SparkHealth {
                ollama_up: true,
                model_ready: false,
            },
        },
        _ => down,
    }
}

fn friendly_err(e: &reqwest::Error) -> String {
    if e.is_connect() || e.is_timeout() {
        "連不上 Ollama（是否未啟動？）".into()
    } else {
        e.to_string()
    }
}

// ── 解析（純函式,可單測）────────────────────────────────────────────────────

/// 把 Ollama 原始輸出解析成 SparkResult。
/// 容忍小模型變異：全形/半形冒號、不同 bullet 記號、4b 偶爾殘留的 `<>` 佔位、
/// 以及保險再 strip 一次 `<think>`（雖然已設 think:false）。
pub fn parse_spark(raw: &str) -> SparkResult {
    let cleaned = strip_think(raw);
    let mut directions = Vec::new();
    let mut musing_lines: Vec<String> = Vec::new();
    let mut in_musing = false;

    for line in cleaned.lines() {
        let t = line.trim();
        if t.is_empty() {
            continue;
        }
        // 「隨想」標頭（兼容簡體「随想」）→ 切到隨想模式,同行冒號後的文字也收。
        if t.starts_with("隨想") || t.starts_with("随想") {
            in_musing = true;
            if let Some(rest) = after_colon(t) {
                if !rest.is_empty() {
                    musing_lines.push(rest.to_string());
                }
            }
            continue;
        }
        if in_musing {
            musing_lines.push(t.to_string());
        } else if let Some(d) = clean_direction(t) {
            // 只收 bullet 行 → 「方向:」標頭、空白自動被忽略。
            directions.push(d);
        }
    }

    SparkResult {
        directions,
        musing: musing_lines.join(""),
    }
}

/// 手動 strip `<think>...</think>`（避免引入 regex crate）。未閉合則丟到結尾。
fn strip_think(s: &str) -> String {
    let mut out = String::new();
    let mut rest = s;
    while let Some(start) = rest.find("<think>") {
        out.push_str(&rest[..start]);
        match rest[start..].find("</think>") {
            Some(end) => rest = &rest[start + end + "</think>".len()..],
            None => return out, // 未閉合,丟棄其後
        }
    }
    out.push_str(rest);
    out
}

fn after_colon(t: &str) -> Option<&str> {
    t.split_once([':', '：']).map(|(_, b)| b.trim())
}

/// bullet 行 → 去掉記號、去掉殘留的 `<>` 佔位,回傳乾淨方向；非 bullet 回 None。
fn clean_direction(t: &str) -> Option<String> {
    let body = ["- ", "-", "•", "*", "・", "‧"]
        .iter()
        .find_map(|p| t.strip_prefix(p))?;
    let s = body.trim().trim_matches(['<', '>', ' ']).trim();
    if s.is_empty() {
        None
    } else {
        Some(s.to_string())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_clean_output() {
        let raw = "方向:\n- 水流與靈感\n- 記憶斷裂\n- 溫度與創意\n隨想:\n洗澡時點子像泡沫,一出水就消失。";
        let r = parse_spark(raw);
        assert_eq!(r.directions, vec!["水流與靈感", "記憶斷裂", "溫度與創意"]);
        assert_eq!(r.musing, "洗澡時點子像泡沫,一出水就消失。");
    }

    #[test]
    fn tolerates_fullwidth_colon_and_spaces() {
        // qwen3 實測會吐全形冒號 + 行尾空白
        let raw = "方向：  \n- 甲  \n- 乙\n- 丙  \n隨想：  \n一段聯想。";
        let r = parse_spark(raw);
        assert_eq!(r.directions, vec!["甲", "乙", "丙"]);
        assert_eq!(r.musing, "一段聯想。");
    }

    #[test]
    fn strips_angle_bracket_placeholders() {
        // 4b 在高 temperature 偶爾把方向包成 <...>
        let raw = "方向:\n- <規則vs本能>\n- <身體記憶>\n- <語言變異>\n隨想:\n語言是肌肉記憶。";
        let r = parse_spark(raw);
        assert_eq!(r.directions, vec!["規則vs本能", "身體記憶", "語言變異"]);
    }

    #[test]
    fn strips_think_block() {
        let raw = "<think>讓我想想…</think>方向:\n- 甲\n隨想:\n短想。";
        let r = parse_spark(raw);
        assert_eq!(r.directions, vec!["甲"]);
        assert_eq!(r.musing, "短想。");
    }

    #[test]
    fn handles_inline_musing_after_colon() {
        let raw = "方向:\n- 甲\n- 乙\n- 丙\n隨想:這是同一行的隨想";
        let r = parse_spark(raw);
        assert_eq!(r.musing, "這是同一行的隨想");
    }

    #[test]
    fn ignores_header_and_blank_lines() {
        let raw = "方向:\n\n- 甲\n\n隨想:\n\n想。";
        let r = parse_spark(raw);
        assert_eq!(r.directions, vec!["甲"]);
        assert_eq!(r.musing, "想。");
    }

    /// 端到端：真連本機 Ollama 走完整 Rust 路徑（reqwest→生成→解析）。
    /// 預設略過,手動跑：`cargo test live_generate -- --ignored --nocapture`
    #[test]
    #[ignore = "需本機 Ollama + qwen3.5:4b"]
    fn live_generate_against_ollama() {
        let r = tauri::async_runtime::block_on(generate(
            "qwen3.5:4b",
            "為什麼人會對沉沒成本如此執著",
            0.7,
        ))
        .expect("generate 應成功");
        eprintln!("directions={:?}\nmusing={}", r.directions, r.musing);
        assert!(!r.directions.is_empty(), "方向不該為空");
        assert!(!r.musing.trim().is_empty(), "隨想不該為空");
    }
}

File: src/spark.ts
// 步2 火花破冰閥：前端直打本地 Ollama，對「某主題 × 某角度」生成發散切角破冰。
// 複用 v1 §2.1 定案 prompt（去 <> 佔位 + 台灣慣用詞彙）。火花輸出 ephemeral：
// 不自動入想法牆，人按「採用」才落地該格（守紅線 #2）；偵測不到 Ollama → 入口隱藏。

export const OLLAMA_URL = "http://localhost:11434";
export const SPARK_MODEL = "qwen3.5:4b"; // 6GB 地板兼預設（Design §3 / v1 §2.1）
export const SPARK_TEMPERATURE = 0.7; // 4b 在 0.9+ 偶吐簡體字 + <> 殘留，壓到 0.7

export interface SparkResult {
  directions: string[];
  musing: string;
}

// v1 §2.1 定案模板，擴成「主題 × 角度」情境。currentIdeas 帶入讓火花接著人已想的繼續歪。
// 2026-05-31 調整：① 用具體範例當格式錨點，避免模型複誦「關鍵詞或切角一」這類佔位字；
// ② 要求每條方向是有畫面的場景／比喻（15～35 字），不要過度壓縮成乾關鍵詞。
export function buildSparkPrompt(
  conceptLabel: string,
  lensTitle: string,
  lensHint: string,
  currentIdeas: string[],
): string {
  const mine = currentIdeas.length > 0 ? currentIdeas.join("、") : "（還沒寫）";
  return [
    "你是發想助手。針對下面這個想法，從指定角度給我激發靈感的延伸。不要完整方案，要發散、可以歪、敢跳。",
    "規則：",
    "- 用繁體中文與台灣慣用詞彙。",
    "- 「方向」給三條，每條是一個有畫面的場景或比喻，約 15～35 字；要具體到能想像，別壓縮成乾巴巴的關鍵詞。",
    "  好範例：「手術室：主刀醫師閉眼靠手溫感知血管走向來旋轉器械」。",
    "  壞範例：「聲音靜電」（太短、沒畫面）。",
    "- 「隨想」一段 50 字以內的聯想短文。",
    "只輸出下面兩段，不要任何開場白、編號標題或結尾說明：",
    "",
    "方向：",
    "- ……",
    "- ……",
    "- ……",
    "隨想：",
    "……",
    "",
    `主題：「${conceptLabel}」`,
    `角度：${lensTitle}（${lensHint}）`,
    `我目前想到的：${mine}`,
  ].join("\n");
}

// 剝 qwen3 thinking 殘留（payload 已設 think:false，這是保險，見 SOP 踩坑）。
export function stripThink(text: string): string {
  return text.replace(/<think>[\s\S]*?<\/think>/g, "").trim();
}

// 模型偶爾複誦的佔位／標籤殘留，或範例骨架（……）→ 不該當成真內容
function isPlaceholder(s: string): boolean {
  if (s.length === 0) return true;
  if (/^[…．.、，\s]+$/.test(s)) return true; // 只剩省略號／標點（範例骨架）
  if (/^[（(].*[）)]$/.test(s)) return true; // 整條被括號包住（如「（第一條）」）
  if (/^(關鍵詞或切角|關鍵詞|切角|方向)[一二三四1234]?$/.test(s)) return true; // 純佔位標籤
  return false;
}

// 去掉模型偶爾複誦的標籤前綴：「關鍵詞或切角一：」「方向二、」等（須帶分隔符才剝，避免誤傷正文）
function stripLabelPrefix(s: string): string {
  return s.replace(/^(關鍵詞或切角|關鍵詞|切角|方向)[一二三四1234]?\s*[:：、.)]\s*/, "").trim();
}

// 解析「方向：- … 隨想：…」格式，容錯：全半形冒號、缺段、不同 bullet、佔位殘留都不炸。
export function parseSparkOutput(raw: string): SparkResult {
  const text = stripThink(raw);
  const directions: string[] = [];
  const musingLines: string[] = [];
  let section: "none" | "dir" | "musing" = "none";

  for (const line of text.split("\n")) {
    const t = line.trim();
    if (t.length === 0) continue;
    if (/^方向[:：]?$/.test(t)) {
      section = "dir";
      continue;
    }
    // 隨想標題：純標題或帶 inline 內容（隨想：xxx）都切到 musing，
    // 否則尾端有字時躲過 header 比對，會被下方裸行分支誤收成方向。
    const musM = t.match(/^隨想[:：]\s*(.*)$/);
    if (musM) {
      section = "musing";
      const rest = musM[1].trim();
      if (rest && !isPlaceholder(rest)) musingLines.push(stripLabelPrefix(rest));
      continue;
    }
    if (section === "dir" && /^[-*・•]/.test(t)) {
      const d = stripLabelPrefix(t.replace(/^[-*・•]\s*/, "").trim());
      if (!isPlaceholder(d)) directions.push(d);
    } else if (section === "dir") {
      // 模型在「方向：」後直接給無 bullet 的裸行也收（段落標題已在上方 continue 濾掉）
      const d = stripLabelPrefix(t);
      if (!isPlaceholder(d)) directions.push(d);
    } else if (section === "musing") {
      if (!isPlaceholder(t)) musingLines.push(stripLabelPrefix(t));
    }
  }

  return { directions, musing: musingLines.join(" ") };
}

// 偵測本地 Ollama 是否可達（同時驗 CORS 有沒有放行）；失敗 → 呼叫端把火花入口降級隱藏。
export async function detectOllama(timeoutMs = 2500): Promise<boolean> {
  try {
    const ctrl = new AbortController();
    const timer = setTimeout(() => ctrl.abort(), timeoutMs);
    try {
      const res = await fetch(`${OLLAMA_URL}/api/tags`, { signal: ctrl.signal });
      return res.ok;
    } finally {
      clearTimeout(timer); // fetch 拋例外時也要清，否則 timer 殘留到 timeoutMs 才觸發
    }
  } catch {
    return false;
  }
}

// 召喚火花：直打 /api/generate。think:false（qwen3 thinking 預設開，會吐 <think> 污染）、stream:false。
export async function summonSpark(
  conceptLabel: string,
  lensTitle: string,
  lensHint: string,
  currentIdeas: string[],
  timeoutMs = 30000,
): Promise<SparkResult> {
  const prompt = buildSparkPrompt(conceptLabel, lensTitle, lensHint, currentIdeas);
  // 本地推論卡死時 fetch 會永不 resolve → loading 解不開；用 AbortController 設逾時上限。
  const ctrl = new AbortController();
  const timer = setTimeout(() => ctrl.abort(), timeoutMs);
  try {
    const res = await fetch(`${OLLAMA_URL}/api/generate`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        model: SPARK_MODEL,
        prompt,
        stream: false,
        think: false,
        options: { temperature: SPARK_TEMPERATURE },
      }),
      signal: ctrl.signal,
    });
    if (!res.ok) {
      throw new Error(`Ollama 回應 ${res.status}`);
    }
    const data = (await res.json()) as { response?: string };
    return parseSparkOutput(data.response ?? "");
  } finally {
    clearTimeout(timer);
  }
}
